src/data-source/collection-browser-data-source.ts

```diff
@@ -0,0 +1,1189 @@
+   */
+  prefixFilterCountMap: Partial<Record<PrefixFilterType, PrefixFilterCounts>> =
+    {};
+
+  constructor(
+    /** The host element to which this controller should attach listeners */
+    private readonly host: ReactiveControllerHost &
+      CollectionBrowserSearchInterface,
+    /** Default size of result pages */
+    private pageSize: number
+  ) {
+    this.host.addController(this as CollectionBrowserDataSourceInterface);
+  }
+
+  /**
+   * @inheritdoc
+   */
+  get size(): number {
+    return this.numTileModels;
+  }
+
+  /**
+   * @inheritdoc
+   */
+  addPage(pageNum: number, pageTiles: TileModel[]): void {
+    this.pages[pageNum] = pageTiles;
+    this.numTileModels += pageTiles.length;
+    this.host.requestUpdate();
+  }
+
+  /**
+   * @inheritdoc
+   */
+  getPage(pageNum: number): TileModel[] {
+    return this.pages[pageNum];
+  }
+
+  /**
+   * @inheritdoc
+   */
+  getAllPages(): Record<string, TileModel[]> {
+    return this.pages;
+  }
+
+  /**
+   * @inheritdoc
+   */
+  hasPage(pageNum: number): boolean {
+    return !!this.pages[pageNum];
+  }
+
+  /**
+   * @inheritdoc
+   */
+  getTileModelAt(index: number): TileModel | undefined {
+    const pageNum = Math.floor(index / this.pageSize) + 1;
+    const indexOnPage = index % this.pageSize;
+    return this.pages[pageNum]?.[indexOnPage];
+  }
+
+  /**
+   * @inheritdoc
+   */
+  setPageSize(pageSize: number): void {
+    this.reset();
+    this.pageSize = pageSize;
+  }
+
+  /**
+   * @inheritdoc
+   */
+  reset(): void {
+    this.pages = {};
+    this.aggregations = {};
+    this.yearHistogramAggregation = undefined;
+    this.pageFetchesInProgress = {};
+
+    this.offset = 0;
+    this.numTileModels = 0;
+    this.totalResults = 0;
+
+    this.host.requestUpdate();
+  }
+
+  /**
+   * @inheritdoc
+   */
+  async handleQueryChange(): Promise<void> {
+    this.reset();
+    await Promise.all([
+      this.doInitialPageFetch(),
+      this.host.suppressFacets ? null : this.fetchFacets(),
+    ]);
+  }
+
+  /**
+   * @inheritdoc
+   */
+  map(
+    callback: (model: TileModel, index: number, array: TileModel[]) => TileModel
+  ): void {
+    this.pages = Object.fromEntries(
+      Object.entries(this.pages).map(([page, tileModels]) => [
+        page,
+        tileModels.map((model, index, array) =>
+          model ? callback(model, index, array) : model
+        ),
+      ])
+    );
+    this.host.requestUpdate();
+  }
+
+  /**
+   * @inheritdoc
+   */
+  checkAllTiles(): void {
+    this.map(model => ({ ...model, checked: true }));
+  }
+
+  /**
+   * @inheritdoc
+   */
+  uncheckAllTiles(): void {
+    this.map(model => ({ ...model, checked: false }));
+  }
+
+  /**
+   * @inheritdoc
+   */
+  removeCheckedTiles(): void {
+    // To make sure our data source remains page-aligned, we will offset our data source by
+    // the number of removed tiles, so that we can just add the offset when the infinite
+    // scroller queries for cell contents.
```

Comment:
ok

Reply:
This piece is migrated over largely unchanged from collection browser, and is one area where we could probably perform a more targeted refactor around page structure. Our approach could use a bit of TLC around sensibly handling pages of data that are smaller than the expected page size (whether due to tiles having been removed, or results omitted on the PPS side).